core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala

```diff
@@ -1197,6 +1213,71 @@ private[history] class FsHistoryProvider(conf: SparkConf, clock: Clock)
     KVUtils.open(newStorePath, metadata)
   }
 
+  private def createHybridStore(
+      dm: HistoryServerDiskManager,
+      appId: String,
+      attempt: AttemptInfoWrapper,
+      metadata: AppStatusStoreMetadata): KVStore = {
+
+    val reader = EventLogFileReader(fs, new Path(logDir, attempt.logPath),
+      attempt.lastIndex)
+    val isCompressed = reader.compressionCodec.isDefined
+
+    val memoryUsage = approximateMemoryUsage(reader.totalSize, isCompressed)
+    if (currentInMemoryStoreUsage.get + memoryUsage > maxInMemoryStoreUsage) {
+      throw new IllegalStateException("Not enough in-memory storage to create hybrid store.")
+    }
+    currentInMemoryStoreUsage.addAndGet(memoryUsage)
+    logInfo(s"Attempt creating hybrid store to parse $appId / ${attempt.info.attemptId}. " +
+      s"Requested ${Utils.bytesToString(memoryUsage)} in-memory storage quota.")
+
+    logInfo(s"Leasing disk manager space for app $appId / ${attempt.info.attemptId}...")
+    val lease = dm.lease(reader.totalSize, isCompressed)
+    val isLeaseRolledBack = new java.util.concurrent.atomic.AtomicBoolean(false)
+    var store: HybridStore = null
+    try {
+      store = new HybridStore()
+      val levelDB = KVUtils.open(lease.tmpPath, metadata)
+      store.setLevelDB(levelDB)
+      store.setCachedQuantileKlass(classOf[CachedQuantile])
+      rebuildAppStore(store, reader, attempt.info.lastUpdated.getTime())
+
+      // Start the background thread to dump data to levelDB when writing to
+      // InMemoryStore is completed.
+      store.switchingToLevelDB(new HybridStore.SwitchingToLevelDBListener {
+        override def onSwitchingToLevelDBSuccess: Unit = {
+          levelDB.close()
+          val newStorePath = lease.commit(appId, attempt.info.attemptId)
+          store.setLevelDB(KVUtils.open(newStorePath, metadata))
+          currentInMemoryStoreUsage.addAndGet(-memoryUsage)
+          logInfo(s"Completely switched to leveldb for app" +
+            s" $appId / ${attempt.info.attemptId}. " +
+            s"Released ${Utils.bytesToString(memoryUsage)} in-memory storage quota.")
+        }
+
+        override def onSwitchingToLevelDBFail(e: Exception): Unit = {
+          logWarning(s"Failed to switch to leveldb for app" +
+          s" $appId / ${attempt.info.attemptId}", e)
+          levelDB.close()
+          if (!isLeaseRolledBack.getAndSet(true)) {
+            lease.rollback()
+          }
+        }
+      })
+
+      store
+    } catch {
+      case e: Exception =>
+        store.close()
+        currentInMemoryStoreUsage.addAndGet(-memoryUsage)
```

Comment:
decrementAndGet here too

Reply:
decrementAndGet() only decrements one, there is no method for decrementAndGet(long n)

Reply:
Ah missed it thanks